source/hp/dof_handler.cc

```diff
@@ -1716,39 +1717,55 @@ namespace hp
     // decide whether we need a sequential or a parallel shared/distributed
     // policy and attach corresponding callback functions dealing with the
     // transfer of active_fe_indices
-    if (dynamic_cast<const parallel::distributed::Triangulation<dim, spacedim>
+    if (dynamic_cast<const parallel::DistributedTriangulationBase<dim, spacedim>
                        *>(&this->get_triangulation()))
       {
         policy = std_cxx14::make_unique<
           internal::DoFHandlerImplementation::Policy::ParallelDistributed<
             DoFHandler<dim, spacedim>>>(*this);
 
-        // repartitioning signals
-        tria_listeners.push_back(
-          this->tria->signals.pre_distributed_repartition.connect([this]() {
-            internal::hp::DoFHandlerImplementation::Implementation::
-              ensure_absence_of_future_fe_indices<dim, spacedim>(*this);
-          }));
-        tria_listeners.push_back(
-          this->tria->signals.pre_distributed_repartition.connect(
-            [this]() { this->pre_distributed_active_fe_index_transfer(); }));
-        tria_listeners.push_back(
-          this->tria->signals.post_distributed_repartition.connect(
-            [this] { this->post_distributed_active_fe_index_transfer(); }));
-
-        // refinement signals
-        tria_listeners.push_back(
-          this->tria->signals.pre_distributed_refinement.connect(
-            [this]() { this->pre_distributed_active_fe_index_transfer(); }));
-        tria_listeners.push_back(
-          this->tria->signals.post_distributed_refinement.connect(
-            [this]() { this->post_distributed_active_fe_index_transfer(); }));
-
-        // serialization signals
-        tria_listeners.push_back(
-          this->tria->signals.post_distributed_save.connect([this]() {
-            this->post_distributed_serialization_of_active_fe_indices();
-          }));
+        // attach signals if the triangulation is a
+        // parallel::distributed::Triangulation
+        // TODO: If possible, change interfaces so that the whole hp::DoFHandler
+        // works for any class derived from
+        // parallel::DistributedTriangulationBase as long as the triangulation
+        // class provides with appropriate functionality (the dof handler should
+        // not depend on how exactly things like solution transfer are
+        // implemented by the triangulation class)
+        if (dynamic_cast<
+              const parallel::distributed::Triangulation<dim, spacedim> *>(
+              &this->get_triangulation()))
```

Comment:
I've been staring at this for a long time now, but can't figure out whether we actually need the `if` here. We want to attach ourselves to signals so that we can ensure certain things, or do certain actions, whenever something happens. I think that is true whether the triangulation in use is a `p::d::T`, or any other distributed triangulation.

In other words, *that* we want to hear about certain events is independent of the triangulation in question. The only place where one might think about doing something different is *in the implementation* of the functions being called by these signals.

So my suggestion would be to remove the `if` here, which implies that the only thing you change is the top-level `if`on line 1719.

Reply:
I've thought about the `if` for quite a while as well. First of all, I agree that it shouldn't be there in a clean design (which is, in my opinion, not the case right now for historical reasons).

My reasoning for putting the `if` there for now was the following: The implementation of the functions attaching to the signals is currently specific to `p::d::T`. So, if these functions are called for anything else than `p::d::T`, one runs into an assertion. Now, if one derives an own distributed triangulation class from `DistributedTriangulationBase` and triggers those signals during refinement/coarsening, and the `hp::DoFHandler` is attached to those, this will abort the program. Of course, if the the `hp::DoFHandler` does not attach to the signals, this then means that one does have to take care of proper assignment of active fe indices, etc. by oneself.

So, as indicated by my TODO comment above the `if` in question, I think the proper approach would be to make the `hp::DoFHandler` completely independent of how exactly things like the transfer of cell data are implemented. But this will require more changes to the triangulation interfaces, which is something I didn't have the time for right now, but which should definitely be done in the future.

But that's just my opinion. If you say, the `if` should go, that's fine for me. In my own application I can just make sure that the signals are never triggered by the triangulation and then I'm good.

Reply:
What I'm realizing just now: What I can of course do, is to remove the assertions in the functions attaching to the signals and just let them do nothing in case it's not a `p::d::T`. That should be consistent with what you suggest. Do you agree?

Reply:
Right. That seems like the better approach. Or, if you want to be on the safe side, remove the `if` here and in the functions that are called do something like
```
  Assert (dynamic_cast<p::d::T*>(...) != nullptr, ExcNotImplemented());
```
so that if the signal is triggered, one gets an error in the place that will have to be checked for any new triangulation class, rather than in an unrelated location.

This also means that a program is fine as long as none of these functions are ever triggered. For example, if you run your program with your own triangulation, then just don't refine (or just don't trigger the signal) and everything will be fine.

Reply:
Ok, I agree and have changed the code accordingly.